File: ThreadApproxPi.java
import java.util.HashSet;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.CountDownLatch;

public class ThreadApproxPi {

    private static final double RADIUS = 1.0;

    private static volatile double totalShots = 0;
    private static volatile double shotsInTarget = 0;

    private static CountDownLatch latch;


    public static void main(String[] args) throws InterruptedException {

        if (args.length != 2) {
            System.out.println("Usage: $ java ApproxPi numberOfTries numberOfThreads");
            System.exit(1);
        }

        int numberOfThreads = Integer.parseInt(args[1]);
        int numberOfTriesForEachThread = Integer.parseInt(args[0]) / numberOfThreads;
        //Set<Thread> threads = new HashSet();
        latch = new CountDownLatch(numberOfThreads);

        //Add new threads
        for (int i = 0; i < numberOfThreads; i++) {
            new Thread(new ApproxPiThread(numberOfTriesForEachThread)).start();
        }



        /*//Start threads
        for (Thread t : threads) {
            t.start();
        }*/

        latch.await();
        System.out.println("Approximation of Pi " + (shotsInTarget / totalShots) * 4);
    }

    private synchronized static void incrementTotalShots (int num) {
        totalShots = totalShots + num;
    }

    private synchronized static void incrementTargetShots (int num) {
        shotsInTarget = shotsInTarget + num;
    }


    @SuppressWarnings("Duplicates")
    static class ApproxPiThread implements Runnable {

        private int numberOfTries;
        private int localTotal;
        private int localTarget;

        public ApproxPiThread(int numberOfTries) {
            this.numberOfTries = numberOfTries;
            this.localTotal = 0;
            this.localTarget = 0;
        }

        @Override
        public void run() {
            try {
                approxPi(numberOfTries);
            } finally {
                latch.countDown();
            }
        }

        private void approxPi(int numberOfTries) {
            Random r = new Random();

            for (int i = 0; i < numberOfTries; i++) {
                double xCoor = generateRandom(r);
                double yCoor = generateRandom(r);

                double distance = calculateDistanceToPoint(xCoor, yCoor);
                localTotal++;

                if (distance <= 1.0) {
                    localTarget++;
                }
            }
            incrementTotalShots(localTotal);
            incrementTargetShots(localTarget);
        }

        private double generateRandom(Random r) {
            int minRange = 0;

            return minRange + (RADIUS - minRange) * r.nextDouble();
        }

        private double calculateDistanceToPoint(double xCoord, double yCoord) {
            return Math.sqrt((xCoord * xCoord) + (yCoord * yCoord));
        }
    }
}

File: ThreadApproxPi.go
package main

import (
	"os"
	"fmt"
	"strconv"
	"sync"
	"math/rand"
	"math"
)

var wg sync.WaitGroup
var mu sync.Mutex

var shotsInTarget = 0.0
var totalShots = 0.0

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Usage: $ go run ApproxPi NumberOfShots NumberOfThreads")
		return
	}

	numberOfThreads := convertToInt(os.Args[2])
	numberOfTriesPerThread := convertToInt(os.Args[1]) / numberOfThreads

	wg.Add(numberOfThreads)

	for i:=0; i<numberOfThreads; i++ {
		go approxPiThread(numberOfTriesPerThread)
	}

	wg.Wait()
	fmt.Println("Approximation of Pi", (shotsInTarget/totalShots)*4)
}

func approxPiThread (numberOfShots int) {
	localTotalShots := 0.0
	localInTarget := 0.0

	for i := 0; i < numberOfShots; i++ {
		xCoord := rand.Float64()
		yCoord := rand.Float64()

		distance := math.Sqrt((xCoord*xCoord)+(yCoord*yCoord))
		localTotalShots++

		if distance <= 1.0 {
			localInTarget++
		}
	}

	setGlobals(localTotalShots, localInTarget)
	wg.Done()
}

func setGlobals (localTotal float64, localInTarget float64) {
	mu.Lock()

	totalShots += localTotal
	shotsInTarget += localInTarget

	mu.Unlock()
}

func convertToInt (stringNumber string) int {
	i,err := strconv.Atoi(stringNumber)
	if err != nil {
		// handle error
		fmt.Println(err)
		os.Exit(2)
	}

	return i
}